security/pkg/nodeagent/cache/helper.go

```diff
@@ -0,0 +1,294 @@
+}
+
+// sendRetriableRequest sends retriable requests for either CSR or ExchangeToken.
+// Prior to sending the request, it also sleep random millisecond to avoid thundering herd problem.
+func (sc *SecretCache) sendRetriableRequest(ctx context.Context, csrPEM []byte, providedExchangedToken, resourceName string, isCSR bool) ([]string, error) {
+	backOffInMilliSec := rand.Int63n(sc.configOptions.InitialBackoff)
+	cacheLog.Debugf("Wait for %d millisec for initial CSR", backOffInMilliSec)
+	// Add a jitter to initial CSR to avoid thundering herd problem.
+	time.Sleep(time.Duration(backOffInMilliSec) * time.Millisecond)
+
+	startTime := time.Now()
+	var retry int64
+	var certChainPEM []string
+	exchangedToken := providedExchangedToken
+	var requestErrorString string
+	var err error
+
+	// Keep trying until no error or timeout.
+	for {
+		var httpRespCode int
+		if isCSR {
+			requestErrorString = fmt.Sprintf("CSR for %q", resourceName)
+			certChainPEM, err = sc.fetcher.CaClient.CSRSign(
+				ctx, csrPEM, exchangedToken, int64(sc.configOptions.SecretTTL.Seconds()))
+		} else {
+			requestErrorString = "Token exchange"
+			p := sc.configOptions.Plugins[0]
+			exchangedToken, _, httpRespCode, err = p.ExchangeToken(ctx, sc.configOptions.TrustDomain, exchangedToken)
+		}
+
+		if err == nil {
+			break
+		}
+
+		// If non-retryable error, fail the request by returning err
+		if !isRetryableErr(status.Code(err), httpRespCode, isCSR) {
+			cacheLog.Errorf("%s hit non-retryable error %v", requestErrorString, err)
+			return nil, err
+		}
+
+		// If reach envoy timeout, fail the request by returning err
+		if startTime.Add(time.Millisecond * envoyDefaultTimeoutInMilliSec).Before(time.Now()) {
+			cacheLog.Errorf("%s retry timed out %v", requestErrorString, err)
+			return nil, err
+		}
+
+		retry++
+		backOffInMilliSec = rand.Int63n(retry * initialBackOffIntervalInMilliSec)
+		time.Sleep(time.Duration(backOffInMilliSec) * time.Millisecond)
+		cacheLog.Warnf("%s failed with error: %v, retry in %d millisec", requestErrorString, err, backOffInMilliSec)
+	}
+
+	if isCSR {
+		return certChainPEM, nil
+	}
+	return []string{exchangedToken}, nil
+}
+
+// getExchangedToken gets the exchanged token for the CSR. The token is either the k8s jwt token of the
+// workload or another token from a plug in provider.
+func (sc *SecretCache) getExchangedToken(ctx context.Context, k8sJwtToken string) (string, error) {
+	exchangedTokens := []string{k8sJwtToken}
+	var err error
+	if sc.configOptions.Plugins != nil && len(sc.configOptions.Plugins) > 0 {
+		// Currently the only plugin is GoogleTokenExchange, so we only extract the first possible plugin.
+		if len(sc.configOptions.Plugins) > 1 {
+			cacheLog.Error("found more than one plugin")
+			return "", err
+		}
+		exchangedTokens, err = sc.sendRetriableRequest(ctx, nil, k8sJwtToken, "", false)
+		if err != nil || len(exchangedTokens) == 0 {
+			cacheLog.Errorf("failed to exchange token: %v", err)
+			return "", err
```

Comment:
could err be nil here ?

Reply:
Let me refine this logic. Actually `len(exchangedTokens) == 0` is always true since I added some code above and forgot to change this line.

Reply:
I removed the check `len(exchangedTokens) == 0` since it will always be false.